packages/fetchai/skills/carpark_client/handlers.py

```diff
@@ -19,15 +19,145 @@
 
 """This package contains the handlers of the agent."""
 
+from typing import Any, cast
+
+from packages.fetchai.protocols.fipa.message import FipaMessage
+from packages.fetchai.protocols.oef_search.message import OefSearchMessage
+from packages.fetchai.skills.generic_buyer.dialogues import (
+    FipaDialogue,
+    FipaDialogues,
+    OefSearchDialogue,
+)
 from packages.fetchai.skills.generic_buyer.handlers import (
     GenericFipaHandler,
     GenericLedgerApiHandler,
     GenericOefSearchHandler,
     GenericSigningHandler,
 )
+from packages.fetchai.skills.generic_buyer.strategy import GenericStrategy
+from packages.fetchai.skills.carpark_client.strategy import Strategy
 
-
-FipaHandler = GenericFipaHandler
 LedgerApiHandler = GenericLedgerApiHandler
-OefSearchHandler = GenericOefSearchHandler
 SigningHandler = GenericSigningHandler
+
+
+class FipaHandler(GenericFipaHandler):
+    """This class handles fipa messages."""
+
+    def __init__(self, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+        self.context.carpark_data = cast(list, [])
+
+    def _handle_propose(
+        self, fipa_msg: FipaMessage, fipa_dialogue: FipaDialogue
+    ) -> None:
+        """
+        Handle the propose.
+
+        :param fipa_msg: the message
+        :param fipa_dialogue: the dialogue object
+        """
+        sender = fipa_msg.sender[-5:]
+        self.context.logger.info(
+            f"received proposal={fipa_msg.proposal.values} from sender={sender}"
+        )
+        strategy = cast(Strategy, self.context.strategy)
+        self.context.carpark_data.append(
+            {
+                "sender": fipa_msg.sender,
+                "message": fipa_msg,
+                "decision": FipaMessage.Performative.DECLINE
+                if not strategy.is_acceptable_proposal(fipa_msg.proposal)
+                or not strategy.is_affordable_proposal(fipa_msg.proposal)
+                else None,
+            }
+        )
+        # self.context.logger.info(
+        #     f"fipa_message_id={fipa_msg.message_id} with target={fipa_msg.target} added to carpark_data."
+        # )
+
+        # makes more sense to only wait for a certain time
```

Comment:
Agree that there should probably be some timeout after which all received proposals are compared. Could also wait for some threshold (less than or equal to the number of agents) to respond. But not sure if that's better than a timeout.

Reply:
I haven't found any solution for non-blocking timeouts in skill handlers so far but if you've got any pointers on how to solve this I'd appreciate it :)

Reply:
Right, so I think this logic would have to be moved over to a new `TickerBehaviour`. Probably you would also need to store a timestamp. Should be doable, just not sure whether it's worth the effort for the demo.

Reply:
Having a new tickerbehaviour did the trick :) thanks guys